ui/src/asm_cleanup.rs

```diff
@@ -107,7 +110,31 @@ pub fn filter_asm(block: &str) -> String {
         }
     }
 
-    let used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    let mut used_labels: HashSet<_> = labels.intersection(&opcode_operands).collect();
+    
+    // We only include labels ref'd by data dirs if the data dir is part of a used label
+    // Check all labels that include data dirs
+    // If label is used, treat all referenced labels from its data dirs as used
+    // Repeat this to check for any addt'l labels found until no more are located
+    let mut checked_labels = HashSet::new();
+    const MAX_ITERATIONS: u32 = 100; // Safeguard against hangs, should not be an issue but better to be safe
+
+    for _ in 0..MAX_ITERATIONS {
+        let mut new_used_label_found = false;
+
+        for (label, references) in &data_dir_label_refs {
+            if used_labels.contains(label) && checked_labels.insert(label) {
+                new_used_label_found = true;
+                
+                for referenced_label in references.iter() {
```

Comment:
Maybe just `used_labels.extend(&references)`?

Reply:
Yes, that's much better.  I was bothered by the nested for loops, but wasn't coming up with anything more expressive.

I decided to just move ```references``` into ```used_labels``` instead of passing a reference, as it's not used again in the HashMap.